In [41]:
import pandas as pd
import numpy as np

In [57]:
pd.set_option('display.max_rows', None)

## Data Catalog

In [2]:
blob = pd.read_excel('Downloads/_Datathon_data_catalog_VIA.xlsx')

In [20]:
blob.columns

Index(['Nr.', 'Name of the data set', 'Files', 'Datathon Category',
       'Organization', 'Data Format', 'Brief description of the data set',
       'Link to access the data ',
       'Additional links to the data / More information'],
      dtype='object')

In [21]:
blob['Link to access the data ']

0         www.viainfo.net/BusService/google_transit.zip
1     http://gtfs.viainfo.net/gtfs-realtime/trapezer...
2     http://gtfs.viainfo.net/tripupdate/tripupdates.pb
3     http://gtfs.viainfo.net/vehicle/vehiclepositio...
4               http://gtfs.viainfo.net/alert/alerts.pb
5               See Datathon Google Cloud Platform Link
6               See Datathon Google Cloud Platform Link
7               See Datathon Google Cloud Platform Link
8               See Datathon Google Cloud Platform Link
9               See Datathon Google Cloud Platform Link
10                                         Coming Soon.
11                                                  NaN
12                                                  NaN
13                                                  NaN
14                                                  NaN
15                                                  NaN
16                                                  NaN
17                                              

In [5]:
blob['Brief description of the data set'].unique()

array(['The General Transit Feed Specification (GTFS), also known as GTFS static or static transit to differentiate it from the GTFS realtime extension, defines a common format for public transportation schedules and associated geographic information. GTFS "feeds" let public transit agencies publish their transit data and developers write applications that consume that data in an interoperable way.',
       'GTFS Realtime is a feed specification that allows public transportation agencies to provide realtime updates about their fleet to application developers. It is an extension to GTFS (General Transit Feed Specification), an open data format for public transportation schedules and associated geographic information. GTFS Realtime was designed around ease of implementation, good GTFS interoperability and a focus on passenger information.',
       nan, 'A sample of schedule adherence grouped by routes',
       'A sample of ridership per route',
       'A sample of early departure, on tim

## Load bus_performance and bus_fare data

In [6]:
bus_performance = pd.read_csv('Downloads/BusOnTimePerformance_2020.csv')

In [9]:
bus_fare = pd.read_excel('Downloads/BusFareBoxActivity_2020.csv')

In [12]:
bus_fare.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89623 entries, 0 to 89622
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   ServiceDateTime      89623 non-null  datetime64[ns]
 1   Route                89623 non-null  int64         
 2   RouteName            87728 non-null  object        
 3   CurrentRevenue       89623 non-null  float64       
 4   Ridership            89623 non-null  int64         
 5   TokenCount           89623 non-null  int64         
 6   TicketCount          89623 non-null  int64         
 7   PassCount            89623 non-null  int64         
 8   BillCount            89623 non-null  int64         
 9   UnclassifiedRevenue  89623 non-null  float64       
 10  DumpCount            89623 non-null  int64         
dtypes: datetime64[ns](1), float64(2), int64(7), object(1)
memory usage: 7.5+ MB


In [27]:
bus_performance.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 107319 entries, 0 to 107318
Data columns (total 10 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   ServiceDateTime   107319 non-null  object 
 1   Route             107319 non-null  int64  
 2   RouteName         107319 non-null  object 
 3   EarlyDeparture    107319 non-null  int64  
 4   OnTime            107319 non-null  int64  
 5   LateArrival       107319 non-null  int64  
 6   Missing           107319 non-null  int64  
 7   TimePointCount    107319 non-null  int64  
 8   OTP               107156 non-null  float64
 9   AverageDwellTime  107155 non-null  float64
dtypes: float64(2), int64(6), object(2)
memory usage: 8.2+ MB


## bus_fare

In [65]:
bus_fare.Route.unique().shape

(225,)

In [66]:
bus_fare.RouteName.unique().shape

(151,)

## bus_performance

In [67]:
bus_performance.Route.unique().shape

(103,)

In [68]:
bus_performance.RouteName.unique().shape

(298,)

#### Needs cleanup if its the truth table for route-route_name

In [29]:
bus_performance.RouteName.unique()

array(['BLANCO', 'SAN PEDRO SKIP', 'SAN PEDRO Frequent', 'McCULLOUGH',
       'US-281 Express', 'STONE OAK/DOWNTOWN Express', 'N ST MARYS',
       'BROADWAY', 'NACO/BROADWAY', 'VIVA CULTURE', 'PERRIN BEITEL SKIP',
       'IH-35 Express', 'NEW BRAUNFELS', 'KIRBY / CONVERSE',
       'HAYS Frequent', 'EAST HOUSTON Frequent', 'E. COMMERCE Frequent',
       'MARTIN LUTHER KING', 'PORTER', 'RIGSBY', 'STEVES AVE.',
       "SOUTH ST.MARY'S", 'S.PRESA', 'VIVA MISSIONS', 'ROOSEVELT',
       'SOUTH FLORES', 'PLEASANTON RD.', 'COMMERCIAL', 'IH-35 S Express',
       'NOGALITOS', 'SOUTH ALAMO', 'KIRK', 'US 90 Express', 'CERALVO',
       'LAREDO', 'GUADALUPE', 'CESAR CHAVEZ', 'W. COMMERCE Frequent',
       'WEST COMMERCE Skip', 'MARTIN', 'RUIZ  Frequent', 'CULEBRA',
       'BANDERA', 'POPLAR', 'WOODLAWN', 'IH-10 CROSSROADS/UTSA Express',
       'IH-10 / FIESTA TEXAS Express', 'FREDERICKSBURG RD',
       'VANCE JACKSON', 'WEST AVE', 'PRIMO FRED RD',
       'UTSA/LEON VALLEY BRT', 'THOUSAND OAKS', 'HUE

## Finding unique pairs of route:routeName in both
#### bus_fare

In [54]:
unique_pairs = bus_fare[['Route', 'RouteName']]

In [60]:
unique_pairs.drop_duplicates().dropna().shape

(150, 2)

#### bus_performance

In [71]:
unique_pairs1 = bus_performance[['Route', 'RouteName']]

In [75]:
unique_pairs1.drop_duplicates().dropna().shape

(300, 2)

In [37]:
len(bus_performance.Route.unique())

103

In [38]:
len(bus_fare.Route.unique())

225

## Find what Route Numbers are in both datasets
I renamed the DFs to fare and performance. Dropped the 'bus' prefix

In [50]:
len(performance[performance.Route.isin(fare.Route)].Route.unique())

102

In [77]:
len(fare[fare.Route.isin(performance.Route)].Route.unique())

102

### 102 both ways, so there is likely 102 routes worth looking at
These are likely to be the route numbers of record

In [86]:
bus_performance.Route.dtype

dtype('int64')

In [79]:
bus_performance.Route.unique()

array([  2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  14,  17,  20,
        21,  22,  24,  25,  26,  28,  30,  32,  34,  36,  40,  42,  43,
        44,  46,  48,  51,  54,  62,  64,  66,  67,  68,  70,  75,  76,
        77,  79,  82,  88,  89,  90,  93,  94,  95,  96,  97, 100, 101,
       502, 503, 505, 509, 515, 520, 522, 524, 534, 550, 551, 602, 603,
       604, 605, 606, 607, 609, 610, 611, 612, 613, 614, 615, 616, 617,
       618, 619, 620, 629, 630, 631, 632, 640, 641, 642, 647, 648, 651,
       660,  16,  65, 999, 705, 712, 715, 103, 501, 672, 102, 552],
      dtype=int64)

truth_table is the route#'s that are in both dataframes

In [80]:
truth_table = pd.DataFrame()

In [81]:
truth_table['Route'] = bus_performance.Route.unique()

best_guess is the unique pairs of route#:routeName with the nicely formated strings in route. 150 of them

In [87]:
best_guess = unique_pairs.drop_duplicates().dropna()

### 97 unique route numbers between those two criteria

In [93]:
best_guess[best_guess.Route.isin(truth_table.Route)].Route.nunique()

97

In [95]:
best_guess[best_guess.Route.isin(truth_table.Route)]

Route                                 RouteName
0          2                              2-BLANCO RD.
1          3                               3-SAN PEDRO
2          4                               4-SAN PEDRO
3          5                             5-MC CULLOUGH
4          8                              8-N.ST.MARYS
5          9                                9-BROADWAY
6         10                           10-NACO./BROAD.
7         11                            11-NACOGDOCHES
8         14                           14-PERRIN BEITL
9         16                        16-Ft. Sam Express
10        17                           17-RANDOLPH P/R
11        20                          20-NEW BRAUNFELS
12        21                                  21-KIRBY
13        22                                  22-NOLAN
14        24                              24-E.HOUSTON
15        25                          25-EAST COMMERCE
16        26                              26-M.L. KING
17        28                                 28-PORTER
18        30                                 30-RIGSBY
19        32                          32-STEVES AVENUE
20        34                            34-S.ST.MARY'S
21        36                           36-S. PRESA ST.
22        40                           40-VIVA Mission
23        42                              42-ROOSEVELT
24        43                              43-S. FLORES
25        44                         44-PLEASANTON RD.
26        46                             46-COMMERCIAL
27        51                       51-NOGALITOS STREET
28        62                           62-KELLY U.S.A.
29        64          64-KLAC/SEA WORLD/N.W. VISTA    
30        65          65-LACKLAND/USO EXPRESS         
31        66                                66-CERALVO
32        67                                 67-LAREDO
33        68                              68-GUADALUPE
34        75                          75-WEST COMMERCE
35        76                         76-OLD HWY. 90 W.
36        77                             77-MARTIN ST.
37        79                               79-RUIZ ST.
38        82                                82-CULEBRA
39        88                            88-BANDERA RD.
40        89                              89-DONALDSON
41        90                            90-BABCOCK RD.
42        94          94-FIESTA/UTSA/EXPRESS          
43        95                     95-Fredericksburg Rd.
44        96                      96-VANCE JACKSON RD.
45        97                            97-WEST AVENUE
46       100             100 - PRIMO FREDRICKSBURG RD.
48       502          502-THOUSAND OAKS               
49       505                             505-BASSE RD.
50       515          515-SOUTHCROSS                  
51       520          520-ZARZAMORA                   
52       522          522-CUPPLES RD.                 
53       524          524-MCMULLEN                    
54       534          534-WURZBACH                    
55       550                   550-LOOPER LIMITED STOP
56       551                   551-LOOPER LIMITED STOP
57       602                            602-MED/N CENT
58       603                       603-MED/WD  PK/UTSA
59       604          604-CROSSROADS/WOODS OF SHAVANO 
60       606          606-INGRAM/NEW TERRITORIES      
61       607                          607-CROSS./INGR.
62       609                            609-BRAUN/INGR
63       610                         610-N.W. CROSSING
64       611                             611-VALLEY HI
65       612                             612-WESTLAKES
66       613                           613-HERITAGE NW
67       614                           614-HIDDEN COVE
68       615                          615-HERITAGE PK.
69       616                           616-SKY HARBOUR
70       617                         617-RAINBOW HILLS
71       618                           618-WLAKES/ING.
72       619                          619-INDIA

# I dont want to write a Regex test to see if these are the ones worth looking at but I bet these 97 are the only ROUTE NUMBERS worth loking at.